lexers/LexTOML.cxx

```diff
@@ -0,0 +1,498 @@
+}
+
+constexpr bool IsNumberContinue(int chPrev, int ch, int chNext) noexcept {
+	return ((ch == '+' || ch == '-') && (chPrev == 'e' || chPrev == 'E'))
+		|| (ch == '.' && chNext != '.');
+}
+
+constexpr bool IsDecimalNumber(int chPrev, int ch, int chNext) noexcept {
+	return IsIdentifierChar(ch) || IsNumberContinue(chPrev, ch, chNext);
+}
+
+constexpr bool IsISODateTime(int ch, int chNext) noexcept {
+	return ((ch == '+' || ch == '-' || ch == ':' || ch == '.') && IsADigit(chNext))
+		|| (ch == ' ' && (chNext == '+' || chNext == '-' || IsADigit(chNext)));
+}
+
+struct EscapeSequence {
+	int outerState = SCE_TOML_DEFAULT;
+	int digitsLeft = 0;
+
+	// highlight any character as escape sequence.
+	bool resetEscapeState(int state, int chNext) noexcept {
+		if (IsEOLChar(chNext)) {
+			return false;
+		}
+		outerState = state;
+		digitsLeft = 1;
+		if (chNext == 'x') {
+			digitsLeft = 3;
+		} else if (chNext == 'u') {
+			digitsLeft = 5;
+		} else if (chNext == 'U') {
+			digitsLeft = 9;
+		}
+		return true;
+	}
+	bool atEscapeEnd(int ch) noexcept {
+		--digitsLeft;
+		return digitsLeft <= 0 || !IsHexDigit(ch);
+	}
+};
+
+constexpr bool IsTripleString(int state) noexcept {
+	return state == SCE_TOML_TRIPLE_STRING_SQ || state == SCE_TOML_TRIPLE_STRING_DQ;
+}
+
+constexpr bool IsDoubleQuoted(int state) noexcept {
+	return state == SCE_TOML_STRING_DQ || state == SCE_TOML_TRIPLE_STRING_DQ;
+}
+
+constexpr int GetStringQuote(int state) noexcept {
+	return IsDoubleQuoted(state) ? '\"' : '\'';
+}
+
+constexpr bool IsTOMLOperator(int ch) noexcept {
+	return AnyOf(ch, '[', ']', '{', '}', ',', '=', '.', '+', '-');
+}
+
+constexpr bool IsTOMLUnquotedKey(int ch) noexcept {
+	return IsIdentifierChar(ch) || ch == '-';
+}
+
+constexpr bool IsWhiteSpace(int ch) noexcept {
+	return (ch == ' ') || ((ch >= 0x09) && (ch <= 0x0d));
+}
+
+static int GetDocNextChar(StyleContext& sc) noexcept {
+	if (!IsWhiteSpace(sc.ch)) {
+		return sc.ch;
+	}
+	if (!IsWhiteSpace(sc.chNext)) {
+		return sc.chNext;
+	}
+	for (Sci_Position pos = 2; ; pos++) {
+		const unsigned char chPos = sc.GetRelative(pos);
+		if (!IsWhiteSpace(chPos)) {
+			return chPos;
+		}
+	}
+}
+
+static int GetLineNextChar(StyleContext& sc) noexcept {
+	if (!IsWhiteSpace(sc.ch)) {
+		return sc.ch;
+	}
+	if (static_cast<Sci_Position>(sc.currentPos) + 1 == sc.lineStartNext) {
+		return '\0';
+	}
+	if (!IsWhiteSpace(sc.chNext)) {
+		return sc.chNext;
+	}
+	for (Sci_Position pos = 2; pos < sc.lineStartNext; pos++) {
+		const unsigned char chPos = sc.GetRelative(pos);
+		if (!IsWhiteSpace(chPos)) {
+			return chPos;
+		}
+	}
+	return '\0';
+}
+
+static bool IsTOMLKey(StyleContext& sc, int braceCount, const WordList *kwList) {
+	if (braceCount) {
+		const int chNext = GetDocNextChar(sc);
+		if (chNext == '=' || chNext == '.' || chNext == '-') {
+			sc.ChangeState(SCE_TOML_KEY);
+			return true;
+		}
+	}
+	if (sc.state == SCE_TOML_IDENTIFIER) {
+		char s[8];
+		sc.GetCurrentLowered(s, sizeof(s));
```

Comment:
@nyamatongwe Just looking for the possible source of the memory corruption and this doesn't seem correct - `GetCurrentLowered()` calls

https://github.com/ScintillaOrg/lexilla/blob/0d49e6cd71a17911f3a63acf27f3b64351edbac3/lexlib/LexAccessor.cxx#L43

which inserts `\0` at `s[8]` in this case. Am I right it should be
```C
sc.GetCurrentLowered(s, sizeof(s) - 1);
```
?

Reply:
`len` is updated with
~~~
endPos_ = std::min(endPos_, startPos_ + len - 1);
len = endPos_ - startPos_;
~~~
which should be a maximum of 1 less than the argument due to the `- 1` in the previous expression. So the largest case is `s[7]` here.

Changing to an unnamed namespace just moved the problem around and it can now be seen as a crash when building with g++ 13.2.0 on Windows.

Reply:
`prev2Type` calculation should be protected against negative line number.
~~~c++
if (lineCurrent >= 2)
	prev2Type = GetLineType(styler.GetLineState(lineCurrent - 2));
~~~

Reply:
`GetLineState()` returns 0 for out-of-range lines.